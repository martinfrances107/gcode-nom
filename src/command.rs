use std::collections::HashSet;

use nom::branch::alt;
use nom::bytes::complete::tag;
use nom::character::complete::digit1;
use nom::combinator::map;
use nom::multi::separated_list1;
use nom::sequence::preceded;
use nom::IResult;

use crate::pos::PosVal;

#[derive(Debug, Clone, PartialEq)]
pub(crate) enum Command<'a> {
    /// linear move
    G1(HashSet<PosVal>),
    /// Home all axes
    G21,
    /// use absolute coordinates.
    G90,
    /// Drop - ie no further action.
    GDrop(&'a str),
    MDrop(&'a str),

    /// No Operation
    Nop,
}

impl<'a> Command<'a> {
    pub(crate) fn parse_line(line: &str) -> IResult<&str, Command> {
        // Most common first.
        alt((
            parse_g1,
            map(tag("G21 "), |_| Command::G21),
            map(tag("G90 "), |_| Command::G90),
            map(g_drop, Command::GDrop),
            map(m_drop, Command::MDrop),
            map(tag(";"), |_| Command::Nop),
            map(tag(""), |_| Command::Nop),
        ))(line)
    }
}

// G commands that require no further action
fn g_drop(i: &str) -> IResult<&str, &str> {
    preceded(tag("G "), digit1)(i)
}

fn parse_g1(i: &str) -> IResult<&str, Command> {
    preceded(
        tag("G1 "),
        map(pos_many, |val: Vec<PosVal>| {
            let mut hs: HashSet<PosVal> = HashSet::new();
            for item in val {
                hs.insert(item);
            }

            Command::G1(hs)
        }),
    )(i)
}

fn pos_many(i: &str) -> IResult<&str, Vec<PosVal>> {
    separated_list1(tag(" "), pos_val)(i)
}

fn pos_val(i: &str) -> IResult<&str, PosVal> {
    alt((
        parse_a, parse_b, parse_c, parse_e, parse_f, parse_g, parse_s, parse_u, parse_v, parse_w,
        parse_x, parse_y, parse_z,
    ))(i)
}

// TODO can I use a macro here!!
fn parse_a(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("A"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::A cannot decode number.");
        PosVal::A(number)
    })(i)
}

fn parse_b(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("B"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::B cannot decode number.");
        PosVal::B(number)
    })(i)
}

fn parse_c(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("C"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::C cannot decode number.");
        PosVal::C(number)
    })(i)
}

fn parse_e(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("E"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::E cannot decode number.");
        PosVal::E(number)
    })(i)
}

fn parse_f(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("F"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::F cannot decode number.");
        PosVal::F(number)
    })(i)
}
fn parse_g(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("G"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::G cannot decode number.");
        PosVal::G(number)
    })(i)
}
// here
fn parse_s(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("S"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::S cannot decode number.");
        PosVal::S(number)
    })(i)
}
fn parse_u(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("U"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::U cannot decode number.");
        PosVal::U(number)
    })(i)
}
fn parse_v(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("V"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::V cannot decode number.");
        PosVal::V(number)
    })(i)
}

fn parse_w(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("W"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::W cannot decode number.");
        PosVal::W(number)
    })(i)
}
fn parse_x(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("X"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::X cannot decode number.");
        PosVal::X(number)
    })(i)
}
fn parse_y(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("Y"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::Y cannot decode number.");
        PosVal::Y(number)
    })(i)
}
fn parse_z(i: &str) -> IResult<&str, PosVal> {
    map(preceded(tag("Z"), digit1), |v: &str| {
        let number = v.parse::<f64>().expect("PosVal::Z cannot decode number.");
        PosVal::Z(number)
    })(i)
}

fn m_drop(i: &str) -> IResult<&str, &str> {
    preceded(tag("M"), digit1)(i)
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn parse_comment() {}

    // #[test]
    // fn many() {
    //     let input = "E5";
    //     let a = pos_many(input);
    //     println!("a {a:#?}");

    //     let b = parse_g1("G1 E5");
    //     println!("b {b:#?}");

    //     let c = parse_g1("G1 E5 F3 ; lift nozzle");
    //     println!("c {c:#?}");
    // }

    #[test]
    fn g1() {
        // let default = PosPayload::<f64>::default();

        let text_commands = [
            ("G1 Z5", Command::G1([PosVal::Z(5_f64)].into())),
            (
                "G1 Z5 F5000 ; lift nozzle",
                Command::G1([PosVal::Z(5_f64), PosVal::F(5000_f64)].into()),
            ),
            // (
            //     "G1 E1.00000 F1800.00000 ; text",
            //     Command::G1([PosVal::E(1.0_f64), PosVal::F(1800_f64)].into()),
            // ),
            // (
            //     "G1 Z0.350 F7800.000",
            //     Command::G1([PosVal::Z(0.350_f64), PosVal::F(7800_f64)].into()),
            // ),
        ];

        for (line, expected) in text_commands {
            let actual = parse_g1(line);
            println!("expected  {expected:?}");
            println!("actual  {actual:?}");
            match actual {
                Ok((_, actual)) => {
                    assert_eq!(actual, expected);
                }
                Err(_) => {
                    assert!(false);
                }
            }
        }
    }
    #[test]
    fn parse_g_drop() {}
}
