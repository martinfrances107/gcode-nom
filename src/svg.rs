use core::fmt::Display;

use crate::{command::Command, pos::PosVal};

#[derive(Debug, Default, Clone)]
pub struct Svg {
    parts: Vec<String>,
}

// A line could not be decoded as an G-Code command
// #[derive(Debug, Clone)]
// struct GCodeError;

// impl std::fmt::Display for GCodeError {
//     fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
//         write!(f, "invalid g-code statement")
//     }
// }

impl Display for Svg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            r#"<svg height="210" width="400" xmlns="http://www.w3.org/2000/svg">
<path d=""#
        )?;
        for part in &self.parts {
            write!(f, "{part}")?;
        }
        write!(
            f,
            r#""
style="fill:none;stroke:green;stroke-width:3" />
 </svg>"#
        )?;
        Ok(())
    }
}

/// Absolute or Relative positioning
#[derive(Default)]
enum CoordPos {
    /// As per spec `CoordPos::Absolute` is the dafault
    /// <https://marlinfw.org/docs/gcode/G090.html>
    #[default]
    Absolute,
    Relative,
}

/// TODO: Want to iterate over something more flexible
/// ie. Drop String for something more generic `AsRef<&str>`?
impl FromIterator<String> for Svg {
    fn from_iter<I>(iter: I) -> Self
    where
        I: IntoIterator<Item = String>,
    {
        let mut svg = Self::default();

        let mut abs_coords = CoordPos::default();
        for line in iter {
            let (_, command) = Command::parse_line(&line).expect("Command not parseable");
            // Error if x,y are missing .. Z default to 0!!
            let mut x = f64::NAN;
            let mut y = f64::NAN;
            let mut z = 0.;
            match command {
                // A non printable move.
                Command::G0(mut payload) => {
                    let params = payload.drain();
                    for param in params {
                        match param {
                            PosVal::X(val) => x = val,
                            PosVal::Y(val) => y = val,
                            PosVal::Z(val) => z = val,
                            PosVal::E(_) | PosVal::F(_) => {
                                // silently drop
                            }
                            pos_bad => {
                                eprintln!("Unexpected param seen in Command::G1 {pos_bad:?}");
                            }
                        }
                    }
                    // Convert x,y,z, into projected x,y.
                    // TODO: Must do something better.
                    let proj_x = x + z / 2.;
                    let proj_y = y + z / 2.;
                    match abs_coords {
                        CoordPos::Absolute => {
                            svg.parts.push(format!("M{proj_x} {proj_y}"));
                        }
                        CoordPos::Relative => {
                            svg.parts.push(format!("m{proj_x} {proj_y}"));
                        }
                    }
                }
                // A printable move.
                Command::G1(mut payload) => {
                    let params = payload.drain();
                    for param in params {
                        match param {
                            PosVal::X(val) => x = val,
                            PosVal::Y(val) => y = val,
                            PosVal::Z(val) => z = val,
                            PosVal::E(_) | PosVal::F(_) => {
                                // silently drop
                            }
                            pos_bad => {
                                eprintln!("Unexpected param seen in Command::G1 {pos_bad:?}");
                            }
                        }
                    }
                    // Convert x,y,z, into projected x,y.
                    // TODO: Must do something better.

                    // Valid `Command::G1` -  Where X and Y and undefined
                    //
                    // "G1 E2.72551 F1800.00000"
                    if !x.is_nan() && !y.is_nan() {
                        let proj_x = x + z / 2.;
                        let proj_y = y + z / 2.;
                        match abs_coords {
                            CoordPos::Absolute => {
                                svg.parts.push(format!("L{proj_x} {proj_y}"));
                            }
                            CoordPos::Relative => {
                                svg.parts.push(format!("l{proj_x} {proj_y}"));
                            }
                        }
                    }
                }
                Command::G21 => svg.parts.push("M0,0".to_string()),
                Command::G90 => abs_coords = CoordPos::Absolute,
                Command::G91 => abs_coords = CoordPos::Relative,
                Command::G92(_) | Command::GDrop(_) | Command::MDrop(_) | Command::Nop => {}
            }
        }

        svg
    }
}
