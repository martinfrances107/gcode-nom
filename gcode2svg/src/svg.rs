use core::fmt::Display;

use gcode_nom::command::Command;
use gcode_nom::parms::PosVal;
use gcode_nom::PositionMode;

#[derive(Debug, Clone)]
pub struct Svg {
    min_x: f64,
    min_y: f64,
    max_x: f64,
    max_y: f64,
    parts: Vec<String>,
}

impl Default for Svg {
    fn default() -> Self {
        Self {
            min_x: f64::INFINITY,
            max_x: -f64::INFINITY,
            min_y: f64::INFINITY,
            max_y: -f64::INFINITY,
            parts: Vec::default(),
        }
    }
}

impl Svg {
    fn update_view_box(&mut self, proj_x: f64, proj_y: f64) {
        // Record min max x, y
        if proj_x > self.max_x {
            self.max_x = proj_x;
        }
        if proj_x < self.min_x {
            self.min_x = proj_x;
        }
        if proj_y > self.max_y {
            self.max_y = proj_y;
        }
        if proj_y < self.min_y {
            self.min_y = proj_y;
        }
    }
}
// A line could not be decoded as an G-Code command
// #[derive(Debug, Clone)]
// struct GCodeError;

// impl std::fmt::Display for GCodeError {
//     fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
//         write!(f, "invalid g-code statement")
//     }
// }

impl Display for Svg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // viewBox="0 0 10 10"
        let width = self.max_x - self.min_x;
        let height = self.max_y - self.min_y;
        let vb = format!("{} {} {} {}", self.min_x, self.min_y, width, height);
        write!(
            f,
            "<svg height=\"{height}\" width=\"{width}\" xmlns=\"http://www.w3.org/2000/svg\" viewBox =\"{vb}\"> <path d=\""
        )?;
        for part in &self.parts {
            write!(f, "{part}")?;
        }
        write!(
            f,
            r#""
style="fill:none;stroke:green;stroke-width:0.05" />
 </svg>"#
        )?;
        Ok(())
    }
}

/// TODO: Want to iterate over something more flexible
/// ie. Drop String for something more generic `AsRef<&str>`?
impl FromIterator<String> for Svg {
    fn from_iter<I>(iter: I) -> Self
    where
        I: IntoIterator<Item = String>,
    {
        let mut svg = Self::default();

        let mut is_extruding = true;
        // Positioning mode for all axes (A, B, C), (U, V, W),  (X, Y, Z).
        let mut position_mode = PositionMode::default();
        // Extruder is unique - can be overridden.
        let mut e_position_mode = PositionMode::default();
        let mut z = 0_f64;
        for line in iter {
            let (_, command) = Command::parse_line(&line).expect("Command not parseable");
            let mut x = f64::NAN;
            let mut y = f64::NAN;

            match command {
                // A non printable move.
                Command::G0(mut payload) => {
                    let params = payload.drain();
                    for param in params {
                        match param {
                            PosVal::X(val) => x = val,
                            PosVal::Y(val) => y = val,
                            PosVal::Z(val) => z = val,
                            PosVal::E(val) => {
                                // Negative values the extruder is "wiping"
                                // or sucking filament back into the extruder.
                                is_extruding = val > 0_f64;
                            }
                            PosVal::F(_) => {
                                // Silently drop.
                            }
                            pos_bad => {
                                eprintln!("Unexpected param seen in Command::G1 {pos_bad:?}");
                            }
                        }
                    }

                    // Valid `Command::G0` -  Where X and Y and undefined
                    //
                    // "G1 E2.72551 F1800.00000"
                    if !x.is_nan() && !y.is_nan() {
                        // Convert x,y,z, into projected x,y.
                        let proj_x = y / 2. + x / 2.;
                        let proj_y = -z - y / 2. + x / 2.;
                        svg.update_view_box(proj_x, proj_y);
                        match position_mode {
                            PositionMode::Absolute => {
                                svg.parts.push(format!("M{proj_x} {proj_y}"));
                            }
                            PositionMode::Relative => {
                                svg.parts.push(format!("m{proj_x} {proj_y}"));
                            }
                        }
                    }
                }
                // A printable move.
                Command::G1(mut params) => {
                    for param in params.drain() {
                        match param {
                            PosVal::X(val) => x = val,
                            PosVal::Y(val) => y = val,
                            PosVal::Z(val) => z = val,
                            PosVal::E(val) => {
                                is_extruding = val > 0_f64;
                            }
                            PosVal::F(_) => {
                                // silently drop
                            }
                            pos_bad => {
                                eprintln!("SVG Unexpected param seen in Command::G1 {pos_bad:?}");
                            }
                        }
                    }
                    // Convert x,y,z, into projected x,y.
                    // TODO: Must do something better.

                    // Valid `Command::G1` -  Where X and Y and undefined
                    //
                    // "G1 E2.72551 F1800.00000"
                    if !x.is_nan() && !y.is_nan() {
                        let proj_x = y / 2. + x / 2.;
                        let proj_y = -z - y / 2. + x / 2.;
                        svg.update_view_box(proj_x, proj_y);
                        match position_mode {
                            PositionMode::Absolute => {
                                if is_extruding {
                                    svg.parts.push(format!("L{proj_x} {proj_y}"));
                                } else {
                                    svg.parts.push(format!("M{proj_x} {proj_y}"));
                                }
                            }
                            PositionMode::Relative => {
                                if is_extruding {
                                    svg.parts.push(format!("l{proj_x} {proj_y}"));
                                } else {
                                    svg.parts.push(format!("m{proj_x} {proj_y}"));
                                }
                            }
                        }
                    }
                }
                Command::G21 => svg.parts.push("M0,0".to_string()),
                Command::G90 => position_mode = PositionMode::Absolute,
                Command::G91 => position_mode = PositionMode::Relative,
                // G92- Set Current Position
                Command::G92(mut params) => {
                    // The extrude rate is going to zero
                    // enter MoveMode ..ie not laying down filament.
                    for param in params.drain() {
                        match param {
                            PosVal::X(val) => x = val,
                            PosVal::Y(val) => y = val,
                            PosVal::Z(val) => z = val,
                            PosVal::E(val) => {
                                // Negative values the extruder is "wiping"
                                // or sucking filament back into the extruder.
                                is_extruding = val > 0_f64;
                            }
                            PosVal::F(_) => {
                                // silently drop
                            }
                            pos_bad => {
                                eprintln!("Unexpected param seen in Command::G1 {pos_bad:?}");
                            }
                        }
                    }

                    // Set Position is by definition a move only.
                    if !x.is_nan() && !y.is_nan() {
                        let proj_x = y / 2. + x / 2.;
                        let proj_y = -z - y / 2. + x / 2.;
                        svg.update_view_box(proj_x, proj_y);
                        match position_mode {
                            PositionMode::Absolute => {
                                svg.parts.push(format!("M{proj_x} {proj_y}"));
                            }
                            PositionMode::Relative => {
                                svg.parts.push(format!("m{proj_x} {proj_y}"));
                            }
                        }
                    }
                }
                Command::GDrop(_) | Command::MDrop(_) | Command::Nop => {}
            }
        }

        svg
    }
}
