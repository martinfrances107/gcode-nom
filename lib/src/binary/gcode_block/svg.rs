use core::fmt::Display;

use crate::command::Command;
use crate::params::PosVal;
use crate::{compute_arc, compute_step_params, ArcParams, PositionMode, StepParams};

/// SVG representation of a G-Code file.
///
/// wraps the min and max x, y values of the SVG.
#[derive(Debug, Clone, PartialEq)]
pub struct Svg {
    min_x: f64,
    min_y: f64,
    max_x: f64,
    max_y: f64,
    parts: Vec<String>,
}

impl Default for Svg {
    fn default() -> Self {
        Self {
            min_x: f64::INFINITY,
            max_x: -f64::INFINITY,
            min_y: f64::INFINITY,
            max_y: -f64::INFINITY,
            parts: Vec::default(),
        }
    }
}

impl Svg {
    fn update_view_box(&mut self, proj_x: f64, proj_y: f64) {
        // Record min max x, y
        if proj_x > self.max_x {
            self.max_x = proj_x;
        }
        if proj_x < self.min_x {
            self.min_x = proj_x;
        }
        if proj_y > self.max_y {
            self.max_y = proj_y;
        }
        if proj_y < self.min_y {
            self.min_y = proj_y;
        }
    }
}
// A line could not be decoded as an G-Code command
// #[derive(Debug, Clone)]
// struct GCodeError;

// impl std::fmt::Display for GCodeError {
//     fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
//         write!(f, "invalid g-code statement")
//     }
// }

impl Display for Svg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let width = self.max_x - self.min_x;
        let height = self.max_y - self.min_y;
        // An empty gcode file will not change min_x/y or max_x/y from
        // its default of +/-INF respectively.
        //
        let parameters = if width.is_finite() && height.is_finite() {
            format!(
                "width=\"{width}\" height=\"{height}\" viewBox=\"{} {} {} {}\"",
                self.min_x, self.min_y, width, height
            )
        } else {
            // In this case silently fail by returning a empty SVG element, without a viewBox parameter.
            String::new()
        };
        writeln!(
            f,
            "<svg xmlns=\"http://www.w3.org/2000/svg\" {parameters} >"
        )?;
        write!(f, "  <path d=\"")?;

        for part in &self.parts {
            write!(f, "{part}")?;
        }
        write!(
            f,
            r#""
style="fill:none;stroke:green;stroke-width:0.05" />
 </svg>"#
        )?;
        Ok(())
    }
}

/// Returns a SVG given a collection of G-Code commands.
///
/// TODO: Want to iterate over something more flexible
/// ie. Drop String for something more generic `AsRef<&str>`?
impl FromIterator<String> for Svg {
    fn from_iter<I>(iter: I) -> Self
    where
        I: IntoIterator<Item = String>,
    {
        let mut svg = Self::default();

        // Invalid if the <path>'s d string does not start with a move.
        svg.parts.push("M0 0".to_string());

        let mut is_extruding = false;
        // Positioning mode for all axes (A, B, C), (U, V, W),  (X, Y, Z).
        let mut position_mode = PositionMode::default();
        let mut z = 0_f64;
        // X and Y position of tool head (before projection).
        let mut current_x = 0_f64;
        let mut current_y = 0_f64;
        for line in iter {
            let (_, command) = Command::parse_line(&line).expect("Command is not parsable");
            // Candidate value of params X<number> Y<number>
            let mut x = f64::NAN;
            let mut y = f64::NAN;

            match command {
                // Treat G0 and G1 command identically.
                //
                // A G0 is a non-printing move but E is present in files seen in the wild.
                // (In the assets directory see the gears and benchy2 files.)
                Command::G0(mut payload) | Command::G1(mut payload) => {
                    // Treat G0 and G1 command identically.
                    //
                    // A G0 is a non-printing move but E is present in files seen in the wild.
                    // (In the assets directory see the gears and benchy2 files.)
                    for param in payload.drain() {
                        match param {
                            PosVal::X(val) => x = val,
                            PosVal::Y(val) => y = val,
                            PosVal::Z(val) => z = val,
                            PosVal::E(val) => {
                                is_extruding = val > 0_f64;
                            }
                            _ => {}
                        }
                    }

                    // Regarding X and Y at least one must be specified.
                    //
                    // If any value is unspecified the current value is used.
                    let xy_valid = match (x.is_nan(), y.is_nan()) {
                        (false, false) => {
                            // X and Y passed as parameters.
                            true
                        }

                        (false, true) => {
                            // X is passed as a parameter
                            // Y is unspecified
                            y = match position_mode {
                                PositionMode::Absolute => current_y,
                                PositionMode::Relative => 0_f64,
                            };
                            true
                        }
                        (true, false) => {
                            // X is unspecified
                            // Y is passed as a parameter
                            x = match position_mode {
                                PositionMode::Absolute => current_x,
                                PositionMode::Relative => 0_f64,
                            };
                            true
                        }

                        (true, true) => {
                            // Cannot proceed: both X and Y are unspecified
                            // Silently handle error by dropping further action.
                            // TODO: Leave a log in the debug output
                            // once debug strategy is worked developed.
                            false
                        }
                    };

                    if xy_valid {
                        let proj_x = y / 2. + x / 2.;
                        let proj_y = -z - y / 2. + x / 2.;
                        svg.update_view_box(proj_x, proj_y);
                        match position_mode {
                            PositionMode::Absolute => {
                                if is_extruding {
                                    svg.parts.push(format!("L{proj_x:.3} {proj_y:.3}"));
                                } else {
                                    svg.parts.push(format!("M{proj_x:.3} {proj_y:.3}"));
                                }
                                current_x = x;
                                current_y = y;
                            }
                            PositionMode::Relative => {
                                if is_extruding {
                                    svg.parts.push(format!("l{proj_x:.3} {proj_y:.3}"));
                                } else {
                                    svg.parts.push(format!("m{proj_x:.3} {proj_y:.3}"));
                                }
                                current_x += x;
                                current_y += y;
                            }
                        }
                    }
                }
                Command::G2(arc_form) => {
                    // Clockwise arc
                    let ArcParams {
                        origin,
                        radius,
                        mut theta_start,
                        theta_end,
                    } = compute_arc(current_x, current_y, &arc_form);

                    // Regarding the Ambiguity/Equivalence  of the angles 0 and 2PI
                    // All values here are in the range 0<=theta<2PI
                    // We are rotating clockwise
                    // in this cased the start angle of 0 should be read as 2PI
                    if theta_start == 0_f64 {
                        theta_start = 2_f64 * std::f64::consts::PI;
                    }

                    let StepParams {
                        n_steps,
                        theta_step,
                    } = compute_step_params(theta_start, theta_end, radius);

                    // For loop: f64 has a problem with numerical accuracy
                    // specifically the comparing limit.
                    // rust idiomatically insists on indexed here
                    for i in 0..=n_steps as u64 {
                        let theta = theta_start + (i as f64 * theta_step);
                        let x = origin.0 + radius * theta.cos();
                        let y = origin.1 + radius * theta.sin();

                        let proj_x = y / 2. + x / 2.;
                        let proj_y = -z - y / 2. + x / 2.;
                        svg.update_view_box(proj_x, proj_y);
                        match position_mode {
                            PositionMode::Absolute => {
                                svg.parts.push(format!("L{proj_x:.3} {proj_y:.3}"));
                            }
                            PositionMode::Relative => {
                                svg.parts.push(format!("l{proj_x:.3} {proj_y:.3}"));
                            }
                        }
                    }
                }
                Command::G3(arc_form) => {
                    // Anti-Clockwise arc
                    let ArcParams {
                        origin,
                        radius,
                        theta_start,
                        mut theta_end,
                    } = compute_arc(current_x, current_y, &arc_form);

                    // Regarding the Ambiguity/Equivalence  of the angles 0 and 2PI
                    // All values here are in the range 0<=theta<2PI
                    // We are rotating anticlockwise
                    // in this cased the final angle of 0 should be read as 2PI
                    if theta_end == 0_f64 {
                        theta_end = 2_f64 * std::f64::consts::PI;
                    }

                    let StepParams {
                        n_steps,
                        theta_step,
                    } = compute_step_params(theta_start, theta_end, radius);

                    // For loop with f64 have a problem with numerical accuracy
                    // specifically the comparing limit.
                    // rust idiomatically insists on indexed here
                    for i in 0..=n_steps as u64 {
                        let theta = theta_start + (i as f64 * theta_step);
                        let x = origin.0 + radius * theta.cos();
                        let y = origin.1 + radius * theta.sin();

                        let proj_x = y / 2. + x / 2.;
                        let proj_y = -z - y / 2. + x / 2.;
                        svg.update_view_box(proj_x, proj_y);
                        match position_mode {
                            PositionMode::Absolute => {
                                svg.parts.push(format!("L{proj_x:.3} {proj_y:.3}"));
                            }
                            PositionMode::Relative => {
                                svg.parts.push(format!("l{proj_x:.3} {proj_y:.3}"));
                            }
                        }
                    }
                }
                Command::G21 => svg.parts.push("M0 0".to_string()),
                Command::G90 => position_mode = PositionMode::Absolute,
                Command::G91 => position_mode = PositionMode::Relative,
                // G92- Set Current Position
                Command::G92(mut params) => {
                    // The extrude rate is going to zero
                    // enter MoveMode ..ie not laying down filament.
                    for param in params.drain() {
                        match param {
                            PosVal::X(val) => x = val,
                            PosVal::Y(val) => y = val,
                            PosVal::Z(val) => z = val,
                            PosVal::E(val) => {
                                // Negative values the extruder is "wiping"
                                // or sucking filament back into the extruder.
                                is_extruding = val > 0_f64;
                            }
                            _ => { /* Silently drop. */ }
                        }
                    }

                    // Set Position is by definition a move only.
                    if !x.is_nan() && !y.is_nan() {
                        let proj_x = y / 2. + x / 2.;
                        let proj_y = -z - y / 2. + x / 2.;
                        svg.update_view_box(proj_x, proj_y);
                        match position_mode {
                            PositionMode::Absolute => {
                                svg.parts.push(format!("M{proj_x} {proj_y}"));
                            }
                            PositionMode::Relative => {
                                svg.parts.push(format!("m{proj_x} {proj_y}"));
                            }
                        }
                    }
                }
                Command::Comment(_) | Command::GDrop(_) | Command::MDrop(_) | Command::Nop => {}
            }
        }

        svg
    }
}

#[cfg(test)]
mod svg {
    use super::*;
    use crate::command::Command;
    use insta::assert_debug_snapshot;

    // The first few lines of assets/3dBench.gcode
    static INPUT: &str = r"
; generated by Slic3r 1.2.9 on 2015-10-01 at 20:51:53

; external perimeters extrusion width = 0.40mm
; perimeters extrusion width = 0.67mm
; infill extrusion width = 0.67mm
; solid infill extrusion width = 0.67mm
; top infill extrusion width = 0.67mm

M107
M190 S65 ; set bed temperature
M104 S205 ; set temperature
G28 ; home all axes
G1 Z5 F5000 ; lift nozzle
M109 S205 ; wait for temperature to be reached
G21 ; set units to millimeters
G90 ; use absolute coordinates
M82 ; use absolute distances for extrusion
G92 E0
G1 E-1.00000 F1800.00000
G92 E0
G1 Z0.350 F7800.000
";

    #[test]
    fn nothing_unhandled() {
        // The first few lines of the benchy file must be recognized.
        for line in INPUT.lines() {
            assert!(Command::parse_line(line).is_ok());
        }
    }

    #[test]
    fn arc_clockwise() {
        // SNAPSHOT tests
        // Simple pattern essential for code coverage
        //
        // Ensures calculated theta values are in the range 0..360
        // as measured in anticlockwise from the positive x-axis.
        //
        // 0 and 360 are the same point
        // This test asserts that the cases where 360 must be used are correct.
        let buffer = include_str!("../../../../assets/g3_box_rounded_anticlockwise.gcode");
        let svg = buffer.lines().map(|l| l.to_string()).collect::<Svg>();
        assert_debug_snapshot!(svg);
    }

    #[test]
    fn arc_anti_clockwise() {
        // SNAPSHOT tests
        // Simple pattern essential for code coverage
        //
        // Ensures calculated theta values are in the range 0..360
        // as measured in anticlockwise from the positive x-axis.
        //
        // 0 and 360 are the same point
        // This test asserts that the cases where 360 must be used are correct.
        let buffer = include_str!("../../../../assets/g2_box_nibble_clockwise.gcode");
        let svg = buffer.lines().map(|l| l.to_string()).collect::<Svg>();
        assert_debug_snapshot!(svg);
    }

    #[test]
    fn arc_demo() {
        // SNAPSHOT tests
        let buffer = include_str!("../../../../assets/arc_demo.gcode");
        let svg = buffer.lines().map(|l| l.to_string()).collect::<Svg>();
        assert_debug_snapshot!(svg);
    }
}
