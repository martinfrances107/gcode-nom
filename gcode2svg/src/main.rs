//! gcode2svg
//!
//! A streaming parser
//!
//! Pass a gcode file into stdin a obj file will be output to StdOut
//!
#![deny(clippy::all)]
#![warn(clippy::cargo)]
#![warn(clippy::complexity)]
#![warn(clippy::pedantic)]
#![warn(clippy::nursery)]
#![warn(clippy::perf)]
#![warn(missing_debug_implementations)]
#![warn(missing_docs)]
#![allow(clippy::many_single_char_names)]
use std::io::stdin;
use std::io::BufRead;
use std::io::Result;

mod svg;

use svg::Svg;

fn main() -> Result<()> {
    let svg = stdin().lock().lines().map(|l| l.unwrap()).collect::<Svg>();
    println!("{svg}");
    Ok(())
}

#[cfg(test)]
mod test {
    use gcode_nom::command::Command;

    // The first few lines of assets/3dBench.gcode
    static INPUT: &str = r"
; generated by Slic3r 1.2.9 on 2015-10-01 at 20:51:53

; external perimeters extrusion width = 0.40mm
; perimeters extrusion width = 0.67mm
; infill extrusion width = 0.67mm
; solid infill extrusion width = 0.67mm
; top infill extrusion width = 0.67mm

M107
M190 S65 ; set bed temperature
M104 S205 ; set temperature
G28 ; home all axes
G1 Z5 F5000 ; lift nozzle
M109 S205 ; wait for temperature to be reached
G21 ; set units to millimeters
G90 ; use absolute coordinates
M82 ; use absolute distances for extrusion
G92 E0
G1 E-1.00000 F1800.00000
G92 E0
G1 Z0.350 F7800.000
";

    #[test]
    fn nothing_unhandled() {
        // The first few lines of the benchy file must be recognized.
        for line in INPUT.lines() {
            assert!(Command::parse_line(line).is_ok());
        }
    }
}
