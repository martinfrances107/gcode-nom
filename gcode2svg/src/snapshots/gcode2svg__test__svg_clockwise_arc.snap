---
source: gcode2svg/src/main.rs
expression: svg
---
Svg {
    min_x: 1.464466094067262,
    min_y: -48.53553390593274,
    max_x: 98.53553390593274,
    max_y: 48.53553390593274,
    parts: [
        "M0 0",
        "M2.500 2.500",
        "L47.500 47.500",
        "L47.500 47.500",
        "L48.647 48.266",
        "L50.000 48.536",
        "L51.353 48.266",
        "L52.500 47.500",
        "L97.500 2.500",
        "L97.500 2.500",
        "L98.266 1.353",
        "L98.536 0.000",
        "L98.266 -1.353",
        "L97.500 -2.500",
        "L52.500 -47.500",
        "L52.500 -47.500",
        "L51.353 -48.266",
        "L50.000 -48.536",
        "L48.647 -48.266",
        "L47.500 -47.500",
        "L2.500 -2.500",
        "L2.500 -2.500",
        "L1.734 -1.353",
        "L1.464 -0.000",
        "L1.734 1.353",
        "L2.500 2.500",
    ],
}
